### DISCUSSION TITLE
Vue 2 compatibility mode

### DISCUSSION THREAD

**Discussion by mikeu** [vue-leaflet/vue-leaflet#158] [Ideas]
This discussion has been a [long time coming](https://github.com/vue-leaflet/vue-leaflet/pull/147) (sorry for the radio silence there, @baptistejamin), but I am hoping to get some feedback from the community at large on the issue of introducing code to support the non-core (though official) `@vue/compat` library, for those working with Vue 2-based projects.

I'll start by saying that I haven't used `compat` myself so might have misunderstood its purpose; if so, this is a great place to elaborate on its intended use to help out the progression of vue-leaflet! My current understanding is that if you have a project written with Vue 2 and would like to upgrade it to Vue 3, you can install this "migration build" and all of your existing Vue 2 code will continue to work as-is, but you'll also be able to start introducing Vue 3 features.

So if your project includes Vue2Leaflet, you could either leave it in place as functioning Vue 2 code until the migration is largely complete and then replace it with vue-leaflet as a final stage, or you could switch to vue-leaflet at any point and it would also run in migration mode, if we include support for it here.

### Pros
* Better support for migrating Vue2Leaflet-based projects to Vue 3 and vue-leaflet.

### Cons
* Introduces code not directly related to Vue or to Leaflet, and that is irrelevant to any new projects started with Vue 3 from the outset.
* Somewhat implies dependence on an additional library—though it does not actually introduce a dependency on `@vue/compat` so far as I can tell.
* Vue2Leaflet already exists for using Leaflet with Vue 2.

At the moment I am leaning toward including this support, but as I said I would appreciate any input that you might have as a member of the vue-leaflet community! Are there any more significant drawbacks to its inclusion that I haven't thought of, or additional motivations for including it? Is there an official recommendation from Vue that all libraries should be adding this if they can? Would it make more sense to introduce a `vue-leaflet-compat` library with this functionality, leaving the core library for Vue 3 only?

**Comment by entioentio** [comment]
Hi there, I'm in the process of migrating a large Vue2 project to Vue3 using compat build. Right now I rely on Vue2leaflet, but at some point I'd love to switch to :point_down: here. I don't have any concrete feedback besides that it's pain to update existing projects. Compat is supposed to make it much easier, but without support for it in the ecosystem, it's not worth much.

**Reply by mikeu** [reply]
Thank you for the feedback, @entioentio, it is useful to hear that support for it would fit into your workflow. I haven't even seen anything from the Vue team suggesting that library authors include the functionality which is a large part of where my uncertainty on including it came from. During your migration process have you found many (any?) of the libraries you use having this support?

**Reply by entioentio** [reply]
Well... No I haven't. I'm using mostly a few component packages. Component framework I was using is the ElementUI, now I'm migrating it to ElementPlus (vue3 - though not vue compat - compatible). I was able to find some alternatives to components I was using before the migration (good example would be vue-multiselect replaced with @vueform/multiselect).

As for support fitting my workflow - I'm pretty confident that the leaflet part of my app will be the last one to migrate, so when the time comes I'll just switch to vue3 production version 🤷. My entry is a bit off-topic. I was expressing my general concern, and since you've asked 😅

**Reply by mikeu** [reply]
Fair enough, and I appreciate that! My thoughts on the compat mode are roughly what you've described I think—first upgrade all of your own code from Vue 2 to Vue 3 using `@vue/compat`, while still using Vue 2 third-party libraries, then remove compat and replace the Vue 2 libraries with their Vue 3 equivalents. In this way the Vue 3-based libraries do not need to know anything about compat mode I think, but since there have been requests for it I've been interested in hearing more about potential alternative approaches.
